rep-0151.rst

```diff
@@ -189,12 +189,108 @@ To support the proposal `B)` less effort is necessary compared to proposal
+disk who would prefer not to have both `python-numpy` and `python3-numpy` installed
+unncessarily.
+
+The following timeline describes a transition over multiple ROS releases:
+
+ * ROS O-turtle (May 2020, LTS):
+
+   * Debian packages using catkin_setup_python install to
+     `lib/python3/dist-packages/`.
+   * From-source builds use Python 3 by default.
+   * All ROS packages should support Python 3.
+
+ * ROS N-turtle (May 2019):
+
+   * Debian packages using catkin_setup_python install to both
+     `lib/python3/dist-packages/` and `lib/python2.7/dist-packages/`
+   * Default behaviour of `catkin_install_python` (with an ambiguous
+     `usr/bin/env python` shebang) is to use Python 3. Users may choose
+     explicitly a `python2` or `python3` shebang, as desired.
+   * From-source builds are using Python 3 by default; users may choose
+     instead Python 2 or to use the dual-build behaviour.
+
+ * ROS Melodic (May 2018, LTS):
+ 
+   * Debian packages using catkin_setup_python install to both
+     `lib/python3/dist-packages/` and `lib/python2.7/dist-packages/`
+   * Default behaviour of `catkin_install_python` (with an ambiguous
+     `usr/bin/env python` shebang) is to use Python 2. Users may choose
+     explicitly a `python2` or `python3` shebang, as desired.
+   * From-source builds are using Python 2 by default; users may choose
+     instead Python 3 or to use the dual-build behaviour.
+
+Requirements
+''''''''''''
+
+To support the proposal `C)` a few capabilities must be developed in catkin:
+
+ * It will need to be updated with bilingual awareness around the
+   `catkin_install_python` and `catkin_python_setup` functions.
+   
+ * The existing `PYTHON_EXECUTABLE`, `PYTHON_LIBRARY`, and `PYTHON_INCLUDE_DIR`
+   variables which may be used today to select the preferred Python will need to
+   be supplemented by a separate `PYTHON_ALT_*` vars which control which Python is
+   used for the second `setup.py` invocation. If the `PYTHON_ALT_*` vars are unset,
+   catkin will build only for the default Python (this would be default behaviour
+   in a source workspace, in order to not break Arch and other from-source systems
+   where `/usr/bin/python` is already Python 3).
+   
+ * Similarly, Python nosetests will run twice if the `PYTHON_ALT_*` vars are set.
+   Perhaps a CMake function could be used to disable this behaviour, especially
+   for packages which migrate to Python 3 and don't want to receive test failures for
+   Python 2 issues. For example, `catkin_python_policy(ONLY_3)` or similar.
+ 
+ * In order to do the right thing with conventional `usr/bin/env python3` shebangs,
+   ROS Melodic will need to ship a small wrapper shim which mutates the `PYTHONPATH`
+   variable. This shim would live at `/opt/ros/melodic/bin/python3` and look like
+    ```
+    #!/bin/sh
+    PYTHONPATH=$(echo $PYTHONPATH | sed "s/lib\/python2\.7/lib\/python3/")
+    /usr/bin/python3 $*
```

Comment:
Coming back to this a year later now that we have largely pursued this strategy (at Clearpath). The main unforeseen effect of this wrapper/env-mangling approach is that once roslaunch moves to Python 3, you end up passing a mangled python3 PYTHONPATH down to the invoked processes. To handle this properly when a process is not yet migrated, you need a second `python2` wrapper that mangles the shebang back again.

Alternatively, you can skip the wrapper and instead use `site.py` and `sitecustomize.py` files in your workspace to mutate the `sys.path` at interpreter startup time. This achieves the same end goal without the annoy propagation side effect.